tempesta_fw/sock.c

```diff
@@ -457,6 +457,28 @@ ss_close(struct sock *sk)
 EXPORT_SYMBOL(ss_close);
 
 /*
+ * The functions ss_do_droplink() and ss_droplink() have the same body.
+ * The difference is that ss_do_droplink() is triggered by the kernel
+ * when FIN comes, whereas the ss_droplink() is called explicitly by
+ * Tempesta. It's essential that the body of these functions is called
+ * only once for a given socket. That is enforced by the following:
+ * - The body is protected by the socket lock which ensures that these
+ *   functions don't run concurrently.
+ * - A socket is no longer live after the body is executed by either
+ *   of these functions.
+ * - If ss_droplink() is called first, then ss_do_droplink() is never
+ *   called by the kernel as the socket will be closed.
+ * - If ss_do_droplink() is called first, then ss_droplink() sees that
+ *   the socket is not live and does not execute the body again.
+ *
+ * All of this ensures that there's always an extra reference to the
+ * socket from Tempesta, until Tempesta is done with the connection.
+ *
+ * XXX: The reasoning above may get invalid if Synchronous Sockets
+ * is changed in regards to the design of socket closing procedure.
+ */
```

Comment:
This still doesn't explain the necessity of ```ss_droplink()``` while we have ```ss_close()```. The functions still do very similar things and having both of them looks like broken design.